include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,632 @@
+  template <typename Op>
+  LinearOperator<Range, Domain> &operator=(const Op &op)
+  {
+    return *this = linop(op);
+  }
+
+  /**
+   * Application of the LinearOperator object to a vector @p u of the
+   * Domain space giving a vector @p v of the Range space.
+   */
+  std::function<const void(Range &v, const Domain &u)> vmult;
+
+  /**
+   * Application of the LinearOperator object to a vector @p u of the
+   * Domain space. The result is added to the vector @p v.
+   */
+  std::function<const void(Range &v, const Domain &u)> vmult_add;
+
+  /**
+   * Application of the transpose LinearOperator object to a vector @p u of
+   * the Range space giving a vector @p v of the Domain space.
+   */
+  std::function<const void(Domain &v, const Range &u)> Tvmult;
+
+  /**
+   * Application of the transpose LinearOperator object to a vector @p u of
+   * the Range space. The result is added to the vector @p v.
+   */
+  std::function<const void(Domain &v, const Range &u)> Tvmult_add;
+
+  /**
+   * Creates a new vector of the Range space that is completeley
+   * initialized such that it can be directly used as the destination
+   * parameter in an application of vmult.
+   */
+  std::function<Range(void)> new_range_member;
+
+  /**
+   * Creates a new vector of the Domain space that is completeley
+   * initialized such that it can be directly used as the source parameter
+   * in an application of vmult.
+   */
+  std::function<Domain(void)> new_domain_member;
+
+  /**
+   * Addition with a LinearOperator @p second_op with the same Domain
+   * and Range.
+   */
+  LinearOperator<Range, Domain> &
+  operator+=(const LinearOperator<Range, Domain> &second_op)
+  {
+    *this = *this + second_op;
+    return *this;
+  }
+
+  /**
+   * Subtraction with a LinearOperator @p second_op with the same Domain
+   * and Range.
+   */
+  LinearOperator<Range, Domain> &
+  operator-=(const LinearOperator<Range, Domain> &second_op)
+  {
+    *this = *this - second_op;
+    return *this;
+  }
+
+  /**
+   * Concatenation of the LinearOperator with an endormorphism @p second_op
+   * on the Domain space.
+   */
+  LinearOperator<Range, Domain> &
+  operator*=(const LinearOperator<Domain, Domain> &second_op)
+  {
+    *this = *this * second_op;
+    return *this;
+  }
+};
+
+
+
+/**
+ * Addition of two linear operators @p first_op and @p second_op given
+ * by $(\text{first_op}+\text{second_op})x:=\text{first_op}(x)+\text{second_op}(x)$
+ */
+template <typename Range, typename Domain>
+LinearOperator<Range, Domain>
+operator+(const LinearOperator<Range, Domain> &first_op,
+          const LinearOperator<Range, Domain> &second_op)
+{
+  LinearOperator<Range, Domain> return_op;
+
+  return_op.new_range_member = first_op.new_range_member;
+  return_op.new_domain_member = first_op.new_domain_member;
+
+  // ensure to have valid computation objects by catching first_op and
+  // second_op by value
+
+  return_op.vmult = [first_op, second_op](Range &v, const Domain &u)
```

Comment:
It is mandatory to capture by value. first_op and second_op might be expiring rvalues, such as in <code>auto foo = linop(A) + linop(B);</code>